File: Backend/src/main/java/be/helha/interf_app/Service/FormService.java
package be.helha.interf_app.Service;

import be.helha.interf_app.Model.Form;
import be.helha.interf_app.Repository.FormRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Service class for managing and performing operations related to {@link Form} entities.
 * This class provides methods to save, retrieve, update, and delete forms from the repository.
 */
@Service
public class FormService {

    /**
     * The repository for accessing and performing CRUD operations on {@link Form} data.
     */
    @Autowired
    private FormRepository formRepository;

    /**
     * Service for managing group-related operations, used to validate form associations.
     */
    @Autowired
    private GroupService groupService;

    /**
     * Saves a new form in the repository.
     * Ensures the form ID is unique and the associated group exists before saving.
     *
     * @param form The form to be saved.
     * @return The saved {@link Form}, or {@code null} if the form already exists or the group is invalid.
     */
    public Form saveForm(Form form) {
        if (getFormById(form.getId()).isEmpty() && groupService.getGroupById(form.getId()).isPresent()) {
            return formRepository.save(form);
        }
        return null;
    }

    /**
     * Retrieves all forms stored in the repository.
     *
     * @return A {@link List} of all {@link Form} objects.
     */
    public List<Form> getAllForms() {
        return formRepository.findAll();
    }

    /**
     * Retrieves a form by its ID from the repository.
     *
     * @param id The ID of the form to retrieve.
     * @return An {@link Optional} containing the {@link Form} if found, or an empty {@link Optional} if not.
     */
    public Optional<Form> getFormById(String id) {
        if (id == null) {
            return Optional.empty();
        }
        return formRepository.findById(id);
    }

    /**
     * Retrieves forms associated with a specific group ID.
     *
     * @param idGroup The ID of the group for which forms are retrieved.
     * @return A {@link List} of {@link Form} objects associated with the given group ID.
     */
    public List<Form> getFormByIdGroup(String idGroup) {
        return formRepository.findByIdGroup(idGroup);
    }

    /**
     * Deletes a form by its ID from the repository.
     *
     * @param id The ID of the form to delete.
     */
    public void deleteForm(String id) {
        formRepository.deleteById(id);
    }

    /**
     * Updates an existing form in the repository.
     * Ensures the form exists and the group ID is unchanged before updating.
     *
     * @param form The form to be updated.
     * @return The updated {@link Form}, or {@code null} if the form does not exist or the group ID does not match.
     */
    public Form updateForm(Form form) {
        if (getFormById(form.getId()).isPresent() &&
                getFormById(form.getId()).get().getIdGroup().equals(form.getIdGroup())) {
            return formRepository.save(form);
        }
        return null;
    }
}

File: Frontend/interfApp/src/app/services/form.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class FormService {

  private baseUrl = 'http://localhost:8080/api/forms'; // URL de base pour l'API backend

  constructor(private http: HttpClient) {}

  // Method to save a form (POST)
  saveForm(form: any): Observable<any> {
    return this.http.post<any>(this.baseUrl, form);
  }

  updateForm(form: any): Observable<any> {
    return this.http.put<any>(this.baseUrl, form);
  }

  // method to get all forms (GET)
  getAllForms(): Observable<any[]> {
    return this.http.get<any[]>(this.baseUrl);
  }

  // (Optionnal) method to get all form by its id (GET)
  getFormById(id: string): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/${id}`);
  }

  getFormByIdGroup(id: string): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/group/${id}`);
  }

  // (Optionnal) method to delete a form (DELETE)
  deleteForm(id: string): Observable<any> {
    console.log(id);
    return this.http.delete(`${this.baseUrl}/${id}`);
  }
}
